### PR TITLE
Use pooch for sample data

### PR DESCRIPTION
**Pull request by dstansby**
This re-writes our sample data handling code to use [`pooch`](https://www.fatiando.org/pooch/latest/). This 

- drops a bunch of custom code
- allows each sample data file to be downloaded on demand instead of download all sample data just to use one file

Checklist:
- [x] Tests - I think existing examples test this thoroughly
- [ ] Check narrative docs for any changes that need making
- [ ] Add changelog
- [ ] Add what's new

Fixes https://github.com/sunpy/sunpy/issues/6244

### CODE REVIEW

File: sunpy/data/sample.py [hunk 1]
```diff
@@ -11,27 +11,63 @@
...
+            return file_dict
+        elif value == 'file_list':
+            return file_dict.values()
+    if value not in AVAILABLE_DATA:
+        raise AttributeError(f'{value} not in sample data set:\n{AVAILABLE_DATA}')
+    return POOCH.fetch(_SAMPLE_DATA[value])
+
+
+def download_sample_data(overwrite=False):
+    return [POOCH.fetch(fname) for fname in FILES]
```
Comment (Cadair) [inline sunpy/data/sample.py:67]: It seems pooch has progress bars off by default, I think we should enable them, if someone's on a slow connection accessing the attribute could lead to your whole session stalling at import with no explanation.

File: sunpy/data/sample.py [hunk 2]
```diff
@@ -11,27 +11,63 @@
...
-    __doc__ += f'   * - ``{keyname}``\n     - {Path(filename).name}\n'
+AVAILABLE_DATA = list(_SAMPLE_DATA.keys())
+FILES = list(_SAMPLE_DATA.values())
+REGISTRY = {fname: None for fname in FILES}
+version = 'v1'
+POOCH = pooch.create(
+    # Use the default cache folder for the operating system
+    path=get_sample_data_dir(),
+    # The remote data is on Github
+    base_url="https://github.com/sunpy/sample-data/raw/master/sunpy/{version}/",
```
[the remark this answers is not in the document]
Comment (Cadair) [reply sunpy/data/sample.py:43]: That is potentially problematic. Having two mirrors makes us less reliant on GH and also allows people in countries or networks which might have GH access restricted a better chance of getting the data.

Also, one of our files doesn't like being downloaded from GH with parfive and always falls back to data.sunpy.org. I assume as pooch is using requests underneath it doesn't have this issue?

**Comment by Cadair** [comment]
I am interested in the difference in transfer speed between this (presumably single threaded) download and parfive which chunks the data with simultaneous requests?

**Comment by Cadair** [comment]
More thoughts:

I am not adverse to this change and adopting pooch, but there are ways we could get a number of the advantages in this PR (only  download files when accessed etc) without adopting pooch and continuing to use parfive. Just worth bearing in mind while weighing up the options.

**Comment by dstansby** [comment]
> In fact if we carried on throwing away the mirror support then this PR would basically only need a few lines changing to keep using parfive (pretty much just the calls to `POOCH.fetch` replace them with `parfive.Downloader.simple_download`).

Yeah, that's a great point, and by implementing this PR I also came to the conclusion that we could achieve the "not downloading all the sample data in one go" goal without `pooch`. So I think it makes sense to close this PR, and do the "don't download all sample data in one go using `__getattr__`" bit first.

**Pull request closed without merging**